bc_obps/service/form_builder_service.py

```diff
@@ -0,0 +1,210 @@
+# report_date is mandatory & determines the valid schemas & WCI configuration for the point in time that the report was created
+
+
+def build_schema(config: int, activity: int, source_types: List[str] | List[int], report_date: str) -> str:
+    # Get activity schema
+    try:
+        activity_schema = (
+            ActivityJsonSchema.objects.get(
+                reporting_activity_id=activity, valid_from__lte=config, valid_to__gte=config
+            )
+        )
+    except:
+        raise Exception(
+            f'No schema found for activity_id {activity} & report_date {report_date}'
+        )
+    rjsf_schema = activity_schema.json_schema
+    gas_type_map: dict[int, str] = {}
+    methodology_map: dict[int, str] = {}
+
+    # Fetch valid source_type(s) for the activity
+    valid_source_types = (
+        ConfigurationElement.objects.select_related('source_type')
+        .filter(
+            reporting_activity_id=activity, valid_from__lte=config, valid_to__gte=config
+        )
+        .distinct('source_type__id')
+    )
+
+    # Except if no valid source_types are found
+    if valid_source_types.count() == 0:
+        raise Exception(f'No valid source_types found for activity_id {activity} & report_date {report_date}')
+    # If an activity only has one source_type, the source type is mandatory and should be added to the schema
+    elif valid_source_types.count() == 1:
+        rjsf_schema['properties']['sourceTypes'] = {"type": "object", "title": "Source Types", "properties": {}}
+        build_source_type_schema(
+            config, activity, int(valid_source_types[0].source_type_id), report_date, rjsf_schema, gas_type_map, methodology_map
+        )
+
+    # If there are multiple source_types for an activity, the user may choose which ones apply. The IDs of the selected source_types are passed as a list in the parameters & we add those schemas to the activity schema.
+    else:
+        for s in valid_source_types:
+            rjsf_schema['properties'][str_to_camel_case(s.source_type.name)] = {
+                "type": "boolean",
+                "title": s.source_type.name,
+            }
+
+    # If no source_types are passed & there are more than 1 valid source type, only return the activity schema
+    if len(source_types) == 0 and valid_source_types.count() > 1:
+        return json.dumps({"schema": rjsf_schema})
+
+    if valid_source_types.count() > 1:
+        # Create the Source Types object (within which all the selected source_type schemas will be defined)
+        rjsf_schema['properties']['sourceTypes'] = {"type": "object", "title": "Source Types", "properties": {}}
+        # For each selected source_type, add the related schema
+        for source_type in source_types:
+            build_source_type_schema(config, activity, int(source_type), report_date, rjsf_schema, gas_type_map, methodology_map)
```

Comment:
And this would become:
```python
rjsf_schema['properties']['sourceTypes']
    ['properties'][str_to_camel_case(source_type.name)] = build_source_type_schema(...)
```